File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        int n=nums.size();
        int subsets=1<<n;
        std::vector<std::vector<int>> ans;
        for(int num=0;num<subsets;num++)
        {
           std::vector<int> subset;
            for(int i=0;i<n;i++)
            {
                if(num & (1<<i))
                {
                    subset.push_back(nums[i]);
                }
               
            }
            ans.push_back(subset);
           
        }
         return ans;
    }
};

File: 0078-subsets/0078-subsets.java


class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(nums, 0, new ArrayList<>(), result);
        return result;
    }

    private void backtrack(int[] nums, int index, List<Integer> current, List<List<Integer>> result) {
        // Add current subset to result
        result.add(new ArrayList<>(current));

        for (int i = index; i < nums.length; i++) {
            // Include nums[i]
            current.add(nums[i]);

            // Move to the next element
            backtrack(nums, i + 1, current, result);

            // Backtrack: remove last element
            current.remove(current.size() - 1);
        }
    }
}
